src/loader/bgw_launcher.c

```diff
@@ -397,54 +421,97 @@ stopped_db_schedulers_cleanup(HTAB *db_htab)
 	}
 }
 
-/* Actions for message types we could receive off of the bgw_message_queue*/
-static void
-message_start_action(HTAB *db_htab, BgwMessage *message, VirtualTransactionId vxid)
+/* Returns the hashtable entry for a database's scheduler, allocating one if it does not exist. */
+static DbHashEntry *
+allocate_scheduler(HTAB *db_htab, Oid db_oid)
 {
-	DbHashEntry *db_he;
 	bool		found;
-	bool		worker_registered;
-	pid_t		worker_pid;
-
-	db_he = hash_search(db_htab, &message->db_oid, HASH_ENTER, &found);
+	DbHashEntry *db_he;
 
-	/*
-	 * This should be idempotent, so if we find the background worker and it's
-	 * not stopped, we should just continue. If we have a worker that stopped
-	 * but hasn't been cleaned up yet, simply restart the worker without
-	 * incrementing cause we've already incremented
-	 */
+	db_he = hash_search(db_htab, &db_oid, HASH_FIND, &found);
 	if (!found)
 	{
+
+		/* Reserve a spot for this scheduler with BGW counter */
 		if (!bgw_total_workers_increment())
 		{
 			report_bgw_limit_exceeded();
-			bgw_message_send_ack(message, ACK_FAILURE);
-			return;
+			return NULL;
 		}
+
+		/*
+		 * Only insert into the hash table if we successfully get a counter
+		 * spot
+		 */
+		db_he = hash_search(db_htab, &db_oid, HASH_ENTER, &found);
+		db_he->db_scheduler_handle = NULL;
 	}
 
-	if (!found || get_background_worker_pid(db_he->db_scheduler_handle, &worker_pid) == BGWH_STOPPED)
+	return db_he;
+}
+
+static bool
+register_scheduler_handle(VirtualTransactionId vxid, DbHashEntry *db_he)
+{
+	pid_t		worker_pid;
+	bool		worker_registered = register_entrypoint_for_db(db_he->db_oid, vxid, &db_he->db_scheduler_handle);
+
+	if (!worker_registered)
 	{
-		worker_registered = register_entrypoint_for_db(db_he->db_oid, vxid, &db_he->db_scheduler_handle);
-		if (!worker_registered)
-		{
-			report_error_on_worker_register_failure();
-			bgw_total_workers_decrement();	/* couldn't register the worker,
-											 * decrement and return false */
-			hash_search(db_htab, &message->db_oid, HASH_REMOVE, &found);
-			bgw_message_send_ack(message, ACK_FAILURE);
-			return;
+		report_error_on_worker_register_failure();
+		return false;
+	}
+	wait_for_background_worker_startup(db_he->db_scheduler_handle, &worker_pid);
+	return true;
+}
 
-		}
-		wait_for_background_worker_startup(db_he->db_scheduler_handle, &worker_pid);
+static bool
+start_or_restart_scheduler(HTAB *db_htab, BgwMessage *message, VirtualTransactionId vxid, bool restart)
+{
+	DbHashEntry *db_he;
+	pid_t		worker_pid;
+
+	db_he = allocate_scheduler(db_htab, message->db_oid);
```

Comment:
Still feels weird to me, and potentially dangerous, to call `allocate_scheduler` on a restart.

It is unclear to me why we are doing this and since you didn't fix this based on my original comment, I assume there is a reason why we're still doing it? If there's a good reason, we should comment on why it makes sense to potentially allocate a new scheduler on a restart (to me it doesn't make semantical sense given the naming, since a restart implicitly implies that a scheduler already exists). Also, why do we even make a distinction between start and restart if we always allocate a scheduler? We could just have a "start" function, which internally checks `found` and first terminates if an existing scheduler was found (and in running state). 

I realize that you might be doing this under the implicit assumption that `allocate_scheduler` always finds an existing scheduler in the restart case, but code-wise it is confusing and doesn't capture the error case when we try to restart a non-existing scheduler.

So, if we make a distinction between a start and a restart, we should for restart only do `HASH_FIND` and fail if not found (not fall back to `HASH_ENTER`). If we don't make a distinction, we should only have a "start" function, and remove the extra message type for the IPC.

Reply:
I think the reason for the confusion here is that we are using restart differently here, restart implies that if there is a scheduler running, it should be stopped and started, whereas you are using it to mean, there is a scheduler running and it should be stopped and started. Start is really, start if not exists or something and restart is stop and start if exists and start if not exists. I'm not sure that there is a better naming convention though, I think these are just different ways of using the words start and restart...(ie restart your computer has a different meaning and you will take different action than if you were told to first start your computer, one would assume that if your computer is already started, you can skip that step, but if told to restart, you can't skip that step. However, if your computer were already shut down, you'd just start it up...). Perhaps a clarifying comment somewhere about how we are using the terms here would make sense? It is clarified in the tests, but having it in the code would make some sense.

Reply:
With these semantics that you describe @davidkohn88, I don't understand why we have both a start and a restart, since the latter would suffice.